e2e/conftest.py

```diff
@@ -0,0 +1,208 @@
+                    break
+                if line is None:
+                    break
+                if line != "":
+                    # logging.warning(f"> {line}")
+                    with self._cond:
+                        self._lines.append(line)
+                        self._cond.notify_all()
+        finally:
+            # If we got here, we're finished reading self._io and need to signal any
+            # waiters that we're done and they'll never hear from us again.
+            with self._cond:
+                self._closed = True
+                self._cond.notify_all()
+
+    def wait_for(self, predicate: Callable[[str], bool], timeoutSecs: float) -> bool:
+        timeoutAt = datetime.datetime.now() + datetime.timedelta(seconds=timeoutSecs)
+        pos = 0
+        with self._cond:
+            while True:
+                while pos < len(self._lines):
+                    if predicate(self._lines[pos]):
+                        return True
+                    pos += 1
+                if self._closed:
+                    return False
+                else:
+                    remaining = (timeoutAt - datetime.datetime.now()).total_seconds()
+                    if remaining < 0 or not self._cond.wait(timeout=remaining):
+                        # Timed out
+                        raise TimeoutError(
+                            "Timeout while waiting for Shiny app to become ready"
+                        )
+
+    def __str__(self):
+        with self._cond:
+            return "".join(self._lines)
+
+
+def dummyio() -> TextIO:
+    io = TextIO()
+    io.close()
+    return io
+
+
+class ShinyAppProc:
+    def __init__(self, proc: subprocess.Popen[str], port: int):
+        self.proc = proc
+        self.port = port
+        self.url = f"http://127.0.0.1:{port}/"
+        self.stdout = OutputStream(proc.stdout or dummyio())
+        self.stderr = OutputStream(proc.stderr or dummyio())
+        threading.Thread(group=None, target=self._run, daemon=True).start()
+
+    def _run(self) -> None:
+        self.proc.wait()
+        if self.proc.stdout is not None:
+            self.proc.stdout.close()
+        if self.proc.stderr is not None:
+            self.proc.stderr.close()
+
+    def close(self) -> None:
+        sleep(0.5)
+        self.proc.terminate()
+
+    def __enter__(self) -> ShinyAppProc:
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ):
+        self.close()
+
+    def wait_until_ready(self, timeoutSecs: float) -> None:
+        if self.stderr.wait_for(
+            lambda line: "Uvicorn running on" in line, timeoutSecs=timeoutSecs
+        ):
+            return
+        else:
+            raise RuntimeError("Shiny app exited without ever becoming ready")
+
+
+def run_shiny_app(
+    app_file: Union[str, PurePath],
+    *,
+    port: Optional[int] = None,
```

Comment:
For consistency with #329...would also require changing line 159 below

```suggestion
    port: int = 0,
```

Reply:
Done!